File: lc0278_FirstBadVersion/lc0278.cc
// LeetCode #278: First Bad Version

#include <iostream>
#include <vector>

using namespace std;

int firstBadVersion (int n) {
    if (n == 0) { return 0; }   // Illegal case.
    if (n == 1) {
        if (isBadVersion (1)) return 1;
        else return 0;
    }
    if (!isBadVersion (1) && !isBadVersion (n)) { return 0; }   // All good versions
    
    int l = 1, r = n;
    while (l < r) {
        int mid = (l + r) / 2;
        if (!isBadVersion (mid)) { l = mid + 1; }
        else { r = mid; }
    }
    return l;
}


int main () {
    
}

File: lc0278_FirstBadVersion/lc0278.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def searchFirstBadVersion(self, l, r):
        """
        :type l: int
        :type r: int
        :rtype: int
        """
        if (l == r):
            return l
        
        m = int((l + r) / 2)        
        
        if (isBadVersion(m)):
            return self.searchFirstBadVersion(l, m)     # We got a one at midpoint. Include it and search the left half
        else:
            return self.searchFirstBadVersion(m+1, r)   # We got a zero at midpoint. Exclude it and search the right half
    
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        if (not isBadVersion(n)):
            return -1   # No bad version at all, if last version is good
        
        f = self.searchFirstBadVersion(1, n)
        
        if (isBadVersion(f)):
            return f
        else:
            return -1   # This should not happen, but for sanity.